### DISCUSSION TITLE
How to code wrapper for envs returning a dictionary of tensors of the observation?

### DISCUSSION THREAD

**Discussion by ErcBunny** [Toni-SM/skrl#185] [Q&A] [answered]
For example my env step returns ``(obs_dict, reward, terminated, truncated, info)`` as type ``Tuple[Dict[str, Tensor], Tensor, Tensor, Tensor, Dict[str, Tensor]]``, and I see the ``wrapper`` base that the step should return ``Tuple[Tensor, Tensor, Tensor, Tensor, Any]``. How should I code the wrapper in this case?

Thanks for reviewing and helping.

**Comment by Toni-SM** [comment]
Hi @ErcBunny 

You can take a look at the Isaac Lab env wrapper that also return a dictionary as observation:

https://github.com/Toni-SM/skrl/blob/636936f3ac49c6d2260bd130d72b789ca6dfe42b/skrl/envs/wrappers/torch/isaaclab_envs.py#L31-L32

**Reply by ErcBunny** [reply]
Thank you for directing me to the isaac lab wrapper, but is _obs_dict["policy"] just a single tensor? I would like to have multiple tensors as observation.

**Reply by Toni-SM** [reply, marked as answer]
Hi @ErcBunny 

Since memory only support tensors (not collection of tensors: dict, tuples, etc.) it is necessary to convert the observation dict (assuming it is a `gymnasium.space.Dict` space) as flattened tensor in the wrapper. Then, in the model, you can map the flattened tensor to the gymnasium space again using the [`Model.tensor_to_space`](https://skrl.readthedocs.io/en/latest/api/models.html#skrl.models.torch.base.Model.tensor_to_space) method.

For converting the space to flattened tensor see the gymnasium wrapper [`_observation_to_tensor`](https://github.com/Toni-SM/skrl/blob/636936f3ac49c6d2260bd130d72b789ca6dfe42b/skrl/envs/wrappers/torch/gymnasium_envs.py#L57) method used in https://github.com/Toni-SM/skrl/blob/636936f3ac49c6d2260bd130d72b789ca6dfe42b/skrl/envs/wrappers/torch/gymnasium_envs.py#L129

Also, see the DeepMind Manipulation: Reach Site Vision [`dm_manipulation_stack_sac.py`](https://github.com/Toni-SM/skrl/blob/main/docs/source/examples/deepmind/dm_manipulation_stack_sac.py) example that show how to use the `tensor_to_space` method.
